### PR TITLE
feat(rustup-mode): install the active toolchain by default on `rustup toolchain install`

### LINKED ISSUE #2686: Command for "install stuff from `rust-toolchain.toml`"
Apologies if this already exists, but I couldn't find it in the docs or shell help. Or if there is an alternate workflow for this.

I love `rust-toolchain.toml`, it's a great idea. It would be even more useful if there were a command such as `rustup override install`, which would look at the current `rust-toolchain.toml` and install the toolchain and components specified there. It would be super useful both when exploring projects for the first time, but also when using disposable container-based development environments.

I'm unopinionated on the exact syntax of this command. Also happy to help implement this if the maintainers want it.

```[tasklist]
### Tasks
- [ ] https://github.com/rust-lang/rustup/issues/3972
```

### PR DESCRIPTION
**Pull request by rami3l**
Closes #2686 by extending the existing `rustup toolchain install` and `rustup install` commands, as discussed in https://github.com/rust-lang/rustup/issues/3635#issuecomment-2260336594:

> > What would be the semantics here? Would it only install if there is a `rust-toolchain.toml` file or would it install any missing default toolchain?
> 
> My understanding is that it is expected to do anything `rustc --version` currently does before actually running that command on the corresponding `rustc`, it's just that everything will become explicit now. So yes, this includes (and isn't limited to) installing the toolchain specified in the `rust-toolchain.toml` file.
> 
> > And presumably the `rustup install` alias.
> 
> Yes! As the two subcommands has been unified with #3596, that looks like a less painful move indeed. A single change for these two places!

TLDR: This PR makes `rustup toolchain install` and `rustup install` install the active toolchain, i.e. behave like the hypothetical `rustup toolchain ensure` command, when no toolchain is specified.

### CODE REVIEW

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
I think conceptually it would be a good next step to break `find_or_install_active_toolchain()` into find and install parts.

**Comment by rami3l** [comment]
> I think conceptually it would be a good next step to break `find_or_install_active_toolchain()` into find and install parts.

@djc The whole story is actually more complex than that. We already have `find_active_toolchain()` but it doesn't equal to `find_or_install_active_toolchain()` minus `ensure_installed()` or something like that. Currently `find_or_install_active_toolchain()` doesn't even call `find_active_toolchain()`, it's a real mess that happens to get this job done.

My plan is that when removing implicit installations altogether, this new call site of that `find_or_install_active_toolchain()` will be the only one, and thus I can start inlining/refactoring it so that all the `verbose` logs and/or sanity checks can hopefully be included.

**Comment by ChrisDenton** [comment]
Is there documentation of the new `rustup toolchain install` behaviour?

**Comment by weihanglo** [comment]
While not really in the same situation, here is the background of why Cargo moved away from implicit `cargo install` to `cargo install --path .`: <https://github.com/rust-lang/cargo/issues/5327>.

**Comment by ChrisDenton** [comment]
Hm, I think rustup is kinda in the exact opposite position. As that issue says, you (almost) never want to `cargo install` the crate in the current directory but you (almost) always want to `rustup install` the toolchain in the crate's rust-toolchain file if you're working on that crate.

**Comment by rami3l** [comment]
> Is there documentation of the new `rustup toolchain install` behaviour?

@ChrisDenton This PR has already changed the `--help` output, would you mind sharing what you have in mind?

**Comment by ChrisDenton** [comment]
I was thinking more of online documentation. Although maybe that should wait until a release. Something along the lines of advice for people using CI. Because presumably we want to guide people towards using this rather than relying on auto install as a side effect of other commands.

**Comment by rami3l** [comment]
> I was thinking more of online documentation. Although maybe that should wait until a release. Something along the lines of advice for people using CI. Because presumably we want to guide people towards using this rather than relying on auto install as a side effect of other commands.

@ChrisDenton The plan is to get rid of implicit installations immediately (#3985), and I promise that will be mentioned in highlight when we cut a new release 🙏
